### DISCUSSION TITLE
[Bug]: imagick extension not loading

### DISCUSSION THREAD

**Discussion by PrasshantChavan** [beyondcode/herd-community#711] [General]
### Platform

Windows

### Operating system version

Windows 11 version 23H2 (0S Build 22631.3527)

### System architecture

Intel (x64) (Windows 11 x64)

### Herd Version

1.1.0

### PHP Version

8.1

### Bug description

As per the herd documentation [https://herd.laravel.com/docs/windows/1/advanced-usage/additional-extensions](url) I have added 
imagick extension but it shows an error log. 

Please help me to solve this error ASAP in herd.

**PHP Warning:  PHP Startup: Unable to load dynamic library 'imagick' (tried: ext\imagick (The specified module could not be found), ext\php_imagick.dll (The specified module could not be found)) in Unknown on line 0**

![2](https://github.com/beyondcode/herd-community/assets/58850768/ac0bf3ee-11cc-4a8a-950b-7d3c52d684d4)
![1](https://github.com/beyondcode/herd-community/assets/58850768/4c7284d3-d31d-4fb5-aaff-d2947c0b5a47)
![3](https://github.com/beyondcode/herd-community/assets/58850768/8c2bf4dc-7663-497e-86d6-9c61e554e92d)

**Comment by mpociot** [comment]
Tr specifying the absolute path to the extension instead.

**Comment by PrasshantChavan** [comment]
Hey @mpociot you mean to say like this

![image](https://github.com/beyondcode/herd-community/assets/58850768/637218c7-aaa5-4e63-a3c6-eaa4069cb12f)

**Comment by mpociot** [comment]
Yeah exactly

**Comment by PrasshantChavan** [comment]
Hey @mpociot ,
Its showing error

**PHP Warning:  PHP Startup: Unable to load dynamic library 'C:\Users\prass\.config\herd\bin\php81\ext\php_imagick.dll' (tried: C:\Users\prass\.config\herd\bin\php81\ext\php_imagick.dll (The specified module could not be found), ext\php_C:\Users\prass\.config\herd\bin\php81\ext\php_imagick.dll.dll (The specified module could not be found)) in Unknown on line 0**

**Comment by nikola-fast** [comment]
Monitoring if someone has found solution, I have same problem. Would be good if there was a tutorial how to install it for Herd.

**Comment by PrasshantChavan** [comment]
Hey @mpociot,
It's a humble request please find out the solution or if it is bug then please fix it or add a feature like installing extensions in specific PHP versions through laravel herd.
Because our working projects are broken because of this issue.
Otherwise, laravel herd is great invention.

Thank you.

**Comment by mpociot** [comment]
In addition to installing the imagemagick PHP extension, you will also need to install and download imagemagick itself: https://stackoverflow.com/a/44259058

**Reply by PrasshantChavan** [reply]
Hey @mpociot ,
I already followed this link when I was using xampp.
But now its not working for herd.

**Comment by msucevan** [comment]
Following, is anyone that for him imagick is working on Windows?

**Comment by nepa-anujpandey** [comment]
Anyone found how to resolve this issue on Windows 11?

**Comment by PrasshantChavan** [comment]
Hey @mpociot 
Please let us know the solution.
Our working projects has been broken

**Comment by pau1phi11ips** [comment]
I had the same problem and it's working in Herd now. I had to install the exe:
https://www.imagemagick.org/script/download.php#windows

and add the DLLs and update php.ini in C:\Users\{my_username}\.config\herd\bin\php83
https://mlocati.github.io/articles/php-windows-imagick.html

**Reply by nilotpaltruepid** [reply]
Have you set any environment variable in windows?

**Comment by msucevan** [comment]
Hi @mpociot, did you have any update on it? Really crucial, because some of my colleagues cannot use some projects with the current version Herd on Windows.

**Reply by pau1phi11ips** [reply]
Ask them to try the steps in my post above https://github.com/beyondcode/herd-community/discussions/711#discussioncomment-10635317

**Comment by akulmambetov** [comment]
@mpociot Any updates?

**Comment by pii-repo** [comment]
@mpociot Any update? I also face the same issue. I am using herd 1.13.0 and php 7.4

**Comment by Ridlo543** [comment]
Ahh, i already fix that in my project, you should download the imagick exactly version, 
donwload in : https://pecl.php.net/package/imagick

check with 
- `php -i|find "Thread Safety"` enabled for thread safe or disabled for not thread safe
- `php -i|findstr "Architecture"` x86 for 32 bits and x64 for 64 bits
- `php -i|find "PHP Version"`

**Comment by petarcar** [comment]
Herd->Services-> click on `i` icon (PHP-*).

```
PHP Warning:  PHP Startup: imagick: Unable to initialize module
Module compiled with build ID=API20230831,TS,VS16
PHP    compiled with build ID=API20230831,NTS,VS16
These options need to match
 in Unknown on line 0
```

There is mismatch post build. None of the versions of imagick suggested above won't work. it's a bug.

**Comment by CyberPunkCodes** [comment]
Just adding some info here:

1, a computer restart seemed to work magic :)
2, I added the path to the imagick folder of all the files from the zip to my windows env path

You need to know what you are working with. Just create a info.php file with `phpinfo()` command and find out what version of PHP (8.4 in my case), whether it's 64 bit or x86, and whether thread safety is enabled or not. Using default Herd for Windows 11, it was 64bit and thread safety is disabled. You can find that all from the php info file.

Then, use this page https://mlocati.github.io/articles/php-windows-imagick.html and click the filters to narrow it down and download the zip file. Open it up. Drag the `php_imagick.dll` to the `ext` directory in your Herd config dir, ie: `C:\Users\[USER]\.config\herd\bin\php84\ext`.

Create a directory in the root of the Herd config dir named `imagick`, ie: `C:\Users\[USER]\.config\herd\bin\php84\imagick`. Extract all of the rest of the files from the zip, except the `php_imagick.dll`, to that directory.

Do a search in your taskbar for `env` and choose the `Edit the system environment..` result. Click the `Environment Variables` button. Click and choose the `Path` result in the top section and click the `Edit` button. Click `New` and add `C:\Users\[USER]\.config\herd\bin\php84\imagick`, replacing your username for your windows system and the version of php if it isnt 8.4. Your first entry should be `C:\Users\[USER]\.config\herd\bin`. Move the imagick one you just created up to be number 2, right below it. Ok, OK, Ok.

In Herd, Go to the PHP section and right click your PHP version, and choose `open php.ini directory`. Edit the `php.ini` file and at the very bottom add the following, assuming you dont have any imagick entries anywhere already:

```bash
[Imagick]
extension="C:\Users\[USER]\.config\herd\bin\php84\ext\php_imagick.dll"
```

Obviously replace your username and the correct version of PHP, it should be the path to the dll file you added first from the zip.

Save and restart your computer.

*DO NOT* install the ImageMagic installer from https://www.imagemagick.org/script/download.php#windows. I was getting the error, "ImageMagick version 1808 but version 1809 is loaded". Uninstalled it, restarted, and it went away and reports both the same version. If you do the above, it isn't needed


Lastly, Restart. It seriously was not working and a restart magically made it work in the end.

Hope that helps someone... or myself in the future :)

**Reply by shahbaz-ali-ssi** [reply]
Just read the first line and restarted system. Solved my issue. I was pulling my hairs over why its working in cli but not showing in herd

**Reply by Hidejec** [reply]
this works for me but it leaves me a message "PHP Warning:  Version warning: Imagick was compiled against ImageMagick version 1808 but version 1810 is loaded. Imagick will run but may behave surprisingly in Unknown on line 0". But it still works as expected.

**Comment by MarJose123** [comment]
This is the solution posted in the Imagick repository:

you can use the latest version of Imagick and make sure your extension is the same with your PHP version

> For those of you like me who have never installed ImageMagick on Windows with PHP, let alone for PHP 7.2, I found this article very useful:
> [Install the ImageMagick PHP extension in Windows](https://mlocati.github.io/articles/php-windows-imagick.html)
> 
> The table here only goes up to PHP 7.1, so I also confirm that using these builds worked for me on PHP 7.2.2/Windows 10 (thread-safe version):
> 
> ImageMagick-7.0.7-11-vc15-x64.zip 
> php_imagick-3.4.3-7.2-ts-vc15-x64.zip
> The workflow seems really clunky to me, but I did exactly as instructed:
> 
> Extract from php_imagick-….zip to the ext directory the php_imagick.dll file of your PHP installation
> Extract from ImageMagick-….zip to the root PHP directory (where php.exe resides) the DLL files that start with (and there are a TON):
> CORE_RL_
> IM_MOD_RL_
> Add this line to your php.ini file
> extension=imagick
> And then to run a simple test (should output a 1):
> php -r "print(class_exists('imagick'));"

[https://github.com/Imagick/imagick/issues/224#issuecomment-367532736](https://github.com/Imagick/imagick/issues/224#issuecomment-367532736)

**Comment by shaffe-fr** [comment]
Hi, I know this post is pretty old, but here's how I installed Imagick on Windows with Herd:

1. Go to https://pecl.php.net/package/imagick and download the latest version for Windows (DLL)
2. Choose the NTS (Non Thread Safe) x64 version matching your PHP version
3. Extract the archive and:
   - Copy all `CORE_*`, `FILTER_*` and `IM_*` files to the same directory as your `php.ini`
   - Copy `php_imagick.dll` to the `ext` folder
4. Add `extension=imagick` to your `php.ini`
5. Restart Herd

To verify: `php -m | grep imagick`
